### PR TITLE
Fix struct row comparator's exception on empty structs

### LINKED ISSUE #10603: [BUG] struct row comparator crashes on empty structs
**Describe the bug**
The latest struct row comparator, which was introduced by #10164, leads to CUDA_ERROR when encountering empty structs.

**Steps/Code to reproduce bug**
```cpp
  std::vector<bool> v{0, 0, 0, 0, 0, 0};
  fixed_width_column_wrapper<int32_t> col_for_mask{{0, 0, 0, 0, 0, 0}, v.begin()};
  auto null_mask  = cudf::copy_bitmask(col_for_mask.release()->view());
  auto struct_col = cudf::make_structs_column(
    6, std::vector<std::unique_ptr<column>>{}, UNKNOWN_NULL_COUNT, std::move(null_mask));

  fixed_width_column_wrapper<int32_t> col1{{1, 2, 3, 1, 2, 3}};
  fixed_width_column_wrapper<int32_t> col2{{1, 1, 1, 2, 2, 2}};
  table_view input{{col1, col2, struct_col->view()}};

  fixed_width_column_wrapper<int32_t> expected{{0, 2, 4, 1, 3, 5}};
  std::vector<order> column_order{order::ASCENDING, order::ASCENDING, order::ASCENDING};

  auto got = sorted_order(input, column_order);
  CUDF_TEST_EXPECT_COLUMNS_EQUAL(expected, got->view());

```

### PR DESCRIPTION
**Pull request by sperlingxx**
Fixes #10603

This PR is to fix a bug of the optimized struct row comparator. For now,  the struct row comparator assumes that structs being compared are non-empty, so it fails when comparing empty structs.

### CODE REVIEW

**Review by devavret — APPROVED** [review]
Looks fine. Any perf hit? You can check using SORT_NVBENCH

**Comment by devavret** [comment]
I didn't know there could be empty structs. Does this work if a struct contains 2 struct children, the first of which is empty?

**Comment by devavret** [comment]
I didn't know there could be empty structs. Does this work if a struct contains 2 struct children, the first of which is empty?

**Comment by jrhemstad** [comment]
> I didn't know there could be empty structs. Does this work if a struct contains 2 struct children, the first of which is empty?

I'm a bit skeptical of that too, though there doesn't seem to be anything in the Arrow spec that forbids it: https://arrow.apache.org/docs/format/Columnar.html#struct-layout

File: cpp/include/cudf/table/experimental/row_operators.cuh [hunk 1]
```diff
@@ -197,7 +197,11 @@ class device_row_comparator {
           return cuda::std::make_pair(state, depth);
         }
 
-        // Structs have been modified to only have 1 child when using this.
+        if (lcol.num_child_columns() == 0) {
+          return cuda::std::make_pair(weak_ordering::EQUIVALENT, depth);
+        }
```
Comment (jrhemstad) [inline cpp/include/cudf/table/experimental/row_operators.cuh:202]: @devavret I assume there is some check earlier on in hose code that verifies corresponding struct columns have the same number of children?

Comment (devavret) [reply cpp/include/cudf/table/experimental/row_operators.cuh:202]: Yes. If it reaches here, they're the same structure.

**Comment by jrhemstad** [comment]
If empty structs always compare equal, then can't we just strip struct columns without children during the pre-processing step?

**Comment by sperlingxx** [comment]
> If empty structs always compare equal, then can't we just strip struct columns without children during the pre-processing step?

Because empty structs may carry null masks, which triggers the null comparsion.

**Review by ttnghia — APPROVED** [review]

**Comment by ttnghia** [comment]
Rerun tests.

**Comment by devavret** [comment]
@sperlingxx mind if I push changes to this? I don't think the changes are sufficient. There is probably another corner case here. https://github.com/rapidsai/cudf/pull/10604#issuecomment-1090177986

**Comment by sperlingxx** [comment]
> SORT_NVBENCH

I added 0 as the depth of struct. Below is the perf result of my local machine: 

```
### [0] NVIDIA TITAN RTX

|     NumRows     | Depth | Nulls | Samples |  CPU Time  | Noise  |  GPU Time  | Noise  |
|-----------------|-------|-------|---------|------------|--------|------------|--------|
|     2^10 = 1024 |     0 |     0 |  11824x |  45.422 us | 11.03% |  42.301 us |  7.77% |
|   2^18 = 262144 |     0 |     0 |   4576x | 112.615 us |  2.82% | 109.586 us |  0.65% |
| 2^26 = 67108864 |     0 |     0 |     44x |  11.388 ms |  0.13% |  11.385 ms |  0.13% |
|     2^10 = 1024 |     1 |     0 |   1630x | 309.766 us |  1.00% | 306.816 us |  0.30% |
|   2^18 = 262144 |     1 |     0 |    499x |   1.006 ms |  0.41% |   1.003 ms |  0.28% |
| 2^26 = 67108864 |     1 |     0 |     11x | 179.348 ms |  0.30% | 179.345 ms |  0.30% |
|     2^10 = 1024 |     8 |     0 |    543x | 924.738 us |  0.37% | 921.886 us |  0.20% |
|   2^18 = 262144 |     8 |     0 |    163x |   3.073 ms |  0.23% |   3.070 ms |  0.21% |
| 2^26 = 67108864 |     8 |     0 |     11x | 589.395 ms |  0.08% | 589.395 ms |  0.08% |
|     2^10 = 1024 |     0 |     1 |   7424x |  70.415 us |  4.55% |  67.393 us |  0.84% |
|   2^18 = 262144 |     0 |     1 |   1716x | 294.552 us |  1.10% | 291.521 us |  0.38% |
| 2^26 = 67108864 |     0 |     1 |     11x |  78.508 ms |  0.09% |  78.505 ms |  0.09% |
|     2^10 = 1024 |     1 |     1 |   1553x | 324.959 us |  1.02% | 322.001 us |  0.43% |
|   2^18 = 262144 |     1 |     1 |   3232x |   1.032 ms | 14.20% |   1.028 ms | 14.20% |
| 2^26 = 67108864 |     1 |     1 |     11x | 180.008 ms |  0.31% | 180.005 ms |  0.31% |
|     2^10 = 1024 |     8 |     1 |   1050x | 940.746 us |  0.60% | 937.648 us |  0.50% |
|   2^18 = 262144 |     8 |     1 |    162x |   3.105 ms |  0.23% |   3.102 ms |  0.20% |
| 2^26 = 67108864 |     8 |     1 |     11x | 588.961 ms |  0.15% | 588.961 ms |  0.15% |
```

**Comment by devavret** [comment]
> > SORT_NVBENCH
> 
> I added 0 as the depth of struct. Below is the perf result of my local machine:
> 
> ```
> ### [0] NVIDIA TITAN RTX
> 
> |     NumRows     | Depth | Nulls | Samples |  CPU Time  | Noise  |  GPU Time  | Noise  |
> |-----------------|-------|-------|---------|------------|--------|------------|--------|
> | 2^26 = 67108864 |     1 |     0 |     11x | 179.348 ms |  0.30% | 179.345 ms |  0.30% |
> | 2^26 = 67108864 |     8 |     0 |     11x | 589.395 ms |  0.08% | 589.395 ms |  0.08% |
> ```

So a minor perf hit for increasing depth vs old one https://github.com/rapidsai/cudf/pull/10164#issuecomment-1068102865
```
----------------------------------------------------------
Benchmark                                            Time 
----------------------------------------------------------
Sort<false>/unstable/67108864/1/manual_time        174 ms 
Sort<false>/unstable/67108864/8/manual_time        517 ms 
```
Depth 1 time increased from 174 ms to 179 ms and depth 8 increased from 517 ms to 589 ms. Not bad. Correctness is more important.

File: cpp/include/cudf/table/experimental/row_operators.cuh (hunk 1 above)
Comment (sperlingxx) [reply cpp/include/cudf/table/experimental/row_operators.cuh:202]: > I didn't know there could be empty structs. Does this work if a struct contains 2 struct children, the first of which is empty?

It does work. I added a test case to cover this scenario. IIUC, following the rule of `decompose_structs`, the second child will be detached from the root struct.

```
     S0
     / \
  S1  S2
```
=>
```
S0
 |
S1   S2
```

**Comment by sperlingxx** [comment]
> 

Of course not. You are the champion of this area, please go ahead :)

**Comment by sperlingxx** [comment]
> @sperlingxx mind if I push changes to this? I don't think the changes are sufficient. There is probably another corner case here. [#10604 (comment)](https://github.com/rapidsai/cudf/pull/10604#issuecomment-1090177986)


Of course not. You are the champion of this area, please go ahead :)

**Comment by sperlingxx** [comment]
Rerun tests.

File: cpp/tests/sort/sort_test.cpp [hunk 2]
```diff
@@ -804,6 +804,60 @@ TYPED_TEST(FixedPointTestAllReps, FixedPointSortedOrderGather)
...
+  auto got2 = sorted_order(input2, {order::DESCENDING});
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(expected2, got2->view());
+
+  // struct{struct{}, struct{int}}
+  int_col col_for_mask2{{0, 0, 0, 0, 0, 0}, {1, 0, 1, 1, 0, 1}};
+  auto null_mask2 = cudf::copy_bitmask(col_for_mask2.release()->view());
+  std::vector<std::unique_ptr<cudf::column>> child_columns2;
+  auto child_col_1 = cudf::make_structs_column(6, {}, UNKNOWN_NULL_COUNT, std::move(null_mask2));
+  child_columns2.push_back(std::move(child_col_1));
+  int_col col4{{0, 1, 2, 3, 4, 5}};
```
Comment (devavret) [inline cpp/tests/sort/sort_test.cpp:847]: Reverse this and the answer will not be the one you expect. You'd expect 415320 but you'd get 145320.

Comment (devavret) [reply cpp/tests/sort/sort_test.cpp]: Fixed https://github.com/rapidsai/cudf/pull/10604/commits/0ee82bc046fb6bf603b0f52377c986210eca8bd9

File: cpp/tests/sort/sort_test.cpp [hunk 3]
```diff
@@ -853,9 +853,9 @@ TEST_F(SortCornerTest, WithEmptyStructColumn)
     cudf::make_structs_column(6, std::move(child_columns2), 0, rmm::device_buffer{});
   table_view input3{{struct_col3->view()}};
 
-  int_col expected3{{1, 4, 0, 2, 3, 5}};
+  int_col expected3{{4, 1, 5, 3, 2, 0}};
   auto got3 = sorted_order(input3, {order::ASCENDING});
-  CUDF_TEST_EXPECT_COLUMNS_EQUAL(expected3, got3->view());
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(expected3, got3->view(), debug_output_level::ALL_ERRORS);
```
Comment (sperlingxx) [inline cpp/tests/sort/sort_test.cpp:858]: Shall we remove `debug_output_level::ALL_ERRORS`?

Comment (devavret) [reply cpp/tests/sort/sort_test.cpp:858]: sure, why not.

File: cpp/tests/sort/sort_test.cpp (hunk 2 above)
Comment (jrhemstad) [reply cpp/tests/sort/sort_test.cpp]: What's the TL;DR on the change made in [#0ee82bc? ](https://github.com/rapidsai/cudf/pull/10604/commits/0ee82bc046fb6bf603b0f52377c986210eca8bd9)

**Comment by sperlingxx** [comment]
@gpucibot merge

**Pull request merged**